### Instructions
1) Start the first cell with your last name, HW/Lab number, etc. For example, HW2_solutions_basnet.

2) After you finish the HW, click File on the top left and click "download as." You will see many options. Please save the file as .ipynb and .html file.

3) To submit HW solutions, go to our course in Canvas --> Assignments --> HW 2 --> Submit Assignment --> Click "Chose file" --> choose HTML file you saved --> Click another file--> choose ipynb file--submit. Once you submit, download  and check to make sure that you have submitted the correct files.
##### ( Both HTML and ipynb files are required for your HW to be graded from now on)

4) Your submission will be compared against other student's solutions and online databases to check plagiarism.

5) You can use this ipynb file to start your solution

6) I will make HWs due on Tuesdays so that you do not have 2 assignments due on Fridays. Each problem is worth 10 points.

In [8]:
import pandas as pd
import numpy as np
coluns2use = ['author_average_rating', 'author_gender','author_name','birthplace','book_average_rating',
              'book_title', 'genre_1','num_ratings', 'num_reviews', 'pages','score']

### Problem 1
Upload the grbooks.csv data in your Jupyter Notebook.  Then, select only the variables listed in the list columns2use in the previous cell and create a new data frame. Find the data type of each of the variables. Do all the variables have correct labels as objects (Categorical) or numbers? Which variable has a wrong datatype?

In [9]:
import pandas as pd
import numpy as np

In [10]:
books = pd.read_csv('grbooks.csv')

In [11]:
books1 = pd.DataFrame({'author_average_rating':books['author_average_rating'], 'author_gender':books['author_gender'], 
                          'author_name':books['author_name'], 'birthplace':books['birthplace'], 'book_average_rating':books['book_average_rating'],
                           'book_title':books['book_title'], 'genre_1':books['genre_1'], 'num_ratings':books['num_ratings'], 
                           'num_reviews':books['num_reviews'], 'pages':books['pages'], 'score':books['score']})

In [12]:
books1.dtypes

author_average_rating    float64
author_gender             object
author_name               object
birthplace                object
book_average_rating      float64
book_title                object
genre_1                   object
num_ratings                int64
num_reviews                int64
pages                     object
score                      int64
dtype: object

pages does not have the right data type. It is a numerical variable labeled as an object while it should be an int.

### Problem 2
Continuing with the data you obtained in problem 1, find the $\textbf{median}$ of all the numerical variables you saw in problem 1. Write a paragraph commenting on the median of these variables. 

In [13]:
books1.median()

author_average_rating       3.97
book_average_rating         3.96
num_ratings              4403.00
num_reviews               384.00
score                    1727.00
dtype: float64

The median of these values is a value that is in the middle of the data. The middle value of author average rating is 3.97, the middle value of book average value is 3.96, the middle value of number of ratings is 4403.00, etc.

### Problem 3 
Find the $\textbf{mode}$ of all the numerical variables you saw in problem 1. Write a paragraph commenting on the mode of these variables. 

In [14]:
books1.mode(numeric_only=True)

author_average_rating  book_average_rating  num_ratings  num_reviews  score
0                   3.94                 3.96           44           17    200

The mode of these values is a value that appears most often in a variable. The number that appears the most often in author average rating is 3.94, in book average rating is 3.96, in number of ratings is 44, in number of reviews is 17, and in score is 200.

### Problem 4
A distribution (here it is one of the numerical columns) in which the values of mean, median, and mode coincide (i.e., mean = median = mode) is known as a normal distribution (a perfect data set).It has a bell shapped curve when plotted and is also symmetric about the mean. Conversely, when mean, median, and mode values are not equal, the distribution is known as asymmetrical or skewed distribution. In moderately skewed or asymmetrical distribution, a very important relationship exists among these three measures of central tendencies. In such distributions, the distance between the mean and median is about one-third of the mean and mode distance. Karl Pearson, a renowned statistician, expressed this relationship as:
$Mode - 3 \times \,\,median - 2\times \,\, mean =0$. 

![](mcube.png) 


Find Mode - 3 Median + 2 Mean for all numerical columns. Symmetric data is hard to come by. If this number is close to 0 for a column, the column data is close to being symmetric, and if it is far from 0 ( positive or negative), the column data is asymmetric. Which column is closest to being symmetric, and which one is the most asymmetric?

In [15]:
(books1.mode(numeric_only=True))-(3*books1.median(numeric_only=True))-(2*books1.mean(numeric_only=True))

author_average_rating  book_average_rating   num_ratings  num_reviews  \
0             -15.890737           -15.822912 -106531.97558 -5784.509589   

          score  
0 -12767.056966

The column that is most symmetric is book average rating column since the value is closest to 0. The column that is most asymmetric is num ratings since the value is the farthest from 0.

### Problem 5
We can use sample function from pandas to randomly select only few  rows from our data frame. Sampling is very useful when the data is too big and not all of it can be handled in a computer.
More about sample function is explained here.
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html


Using the sample function, sample 1000 data points from the data you created in problem 1 and give it a new name. Now find the means  of this new sampled data and compare with the means of the data you sampled from. Which variable's mean  from this sampled data setis close to the mean of the original data? What does it imply when these two means are very close? 

In [16]:
books1.sample(1000, random_state= 16).head()

author_average_rating author_gender        author_name  \
22536                   4.12          male  Lisa   Thompson\n   
19560                   3.87          male      Matt Phelan\n   
17977                   3.97        female      M. Chandler\n   
17939                   4.37          male     Tsugumi Ohba\n   
1134                    4.03        female   Crystal Spears\n   

                   birthplace  book_average_rating  \
22536    United Kingdom\n                     4.10   
19560                    \n                   3.72   
17977  \n   United States\n                   3.85   
17939               Japan\n                   4.18   
1134        United States\n                   4.21   

                                              book_title         genre_1  \
22536                         \n      The Goldfish Boy\n         Mystery   
19560                    \n      The Storm in the Barn\n  Sequential Art   
17977  \n      Shadow of the Templar: The Morning Star\n         Romance   
17939                Bakuman, Band 1: Traum und Realität  Sequential Art   
1134                           \n      Withstanding Me\n         Romance   

       num_ratings  num_reviews pages  score  
22536         3472          631   320    635  
19560         3409          566   208    950  
17977         1807          225     0   1079  
17939        18228          475   208    397  
1134          8105          471   264   5466

In [17]:
sampled_books=books1.sample(1000, random_state= 16).reset_index(drop=True)
sampled_books.head()

author_average_rating author_gender        author_name  \
0                   4.12          male  Lisa   Thompson\n   
1                   3.87          male      Matt Phelan\n   
2                   3.97        female      M. Chandler\n   
3                   4.37          male     Tsugumi Ohba\n   
4                   4.03        female   Crystal Spears\n   

               birthplace  book_average_rating  \
0    United Kingdom\n                     4.10   
1                    \n                   3.72   
2  \n   United States\n                   3.85   
3               Japan\n                   4.18   
4       United States\n                   4.21   

                                          book_title         genre_1  \
0                         \n      The Goldfish Boy\n         Mystery   
1                    \n      The Storm in the Barn\n  Sequential Art   
2  \n      Shadow of the Templar: The Morning Star\n         Romance   
3                Bakuman, Band 1: Traum und Realität  Sequential Art   
4                          \n      Withstanding Me\n         Romance   

   num_ratings  num_reviews pages  score  
0         3472          631   320    635  
1         3409          566   208    950  
2         1807          225     0   1079  
3        18228          475   208    397  
4         8105          471   264   5466

In [18]:
mean1 = books1.mean()
mean1

author_average_rating        3.960368
book_average_rating          3.951456
num_ratings              46683.487790
num_reviews               2324.754794
score                     3893.028483
dtype: float64

In [19]:
mean2 = sampled_books.mean()
mean2

author_average_rating        3.96817
book_average_rating          3.96087
num_ratings              45022.12900
num_reviews               2456.98200
pages                            inf
score                     3515.20800
dtype: float64

In [20]:
mean1 - mean2

author_average_rating      -0.007802
book_average_rating        -0.009414
num_ratings              1661.358790
num_reviews              -132.227206
pages                            NaN
score                     377.820483
dtype: float64

Variables' means that are close to the means of original data are author average rating and book average rating. This implies that the sample set accurately represents the originial data set.

### Problem 6
Make a frequency table of the variable "genre_1". Your answer should look like a data frame with indices ( with no column name), and two other column names "genre_1" and "frequency". The following Panda functions will be useful: groubby, value_counts, reset_index() and to_frame(). https://stackoverflow.com/questions/39778686/pandas-reset-index-after-groupby-value-counts and https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reset_index.html

In [96]:
frequency = books1['genre_1'].value_counts()

In [97]:
key = frequency.keys()

In [98]:
values = frequency.values

In [101]:
data = pd.DataFrame({'genre_1': key, 'frequency': values})

In [103]:
data.reset_index(drop=True)

genre_1  frequency
0         Romance       3397
1         Fantasy       3397
2     Young Adult       2497
3         Fiction       2172
4      Historical       1399
..            ...        ...
119        How To          1
120  Dark Fantasy          1
121       Medical          1
122   Fan Fiction          1
123       Couture          1

[124 rows x 2 columns]

### Problem 7. 
Using the column  "score", to create a new column called new_score that will have a score out of 5. For this, apply the formula  $\text{new_score} = \dfrac{\text{score}}{\text{max(score)}}\times 5$. Print the description of this new colum using describe function.

In [51]:
newscore = books1['score'] / max(books1['score']) * 5

In [52]:
newscore

0        0.026995
1        0.025590
2        0.038319
3        0.012812
4        0.017592
           ...   
22886    0.010154
22887    0.011040
22888    0.014032
22889    0.015854
22890    0.027162
Name: score, Length: 22891, dtype: float64

In [59]:
books1['new_score'] = newscore

### Problem 8.
We have now 3 columns that represent ratings out of 5, namely author_average_rating, book_average_rating, and a new_score we just created in problem 7. Let's create a column called super_score which takes into consideration of all the 3 columns with rating scores out of 5. Give 24%, 34% and 42% weights for the columns author_average_rating, book_average_rating, and a new_score respectively to create the values for super_score column that is also out of 5. Now use the describe function to show the details of this new column.

In [61]:
superscore = .24*books1['author_average_rating'] + .34*books1['book_average_rating'] + .42*books1['new_score']

In [62]:
superscore

0        2.340538
1        2.410948
2        2.332694
3        2.344181
4        2.354789
           ...   
22886    2.248665
22887    2.584637
22888    2.234093
22889    2.337259
22890    2.542008
Length: 22891, dtype: float64

In [63]:
books1['super_score'] = superscore

In [66]:
books1.describe()

author_average_rating  book_average_rating   num_ratings  \
count           22891.000000         22891.000000  2.289100e+04   
mean                3.960368             3.951456  4.668349e+04   
std                 0.240421             0.291317  1.800698e+05   
min                 1.820000             0.000000  0.000000e+00   
25%                 3.810000             3.770000  8.200000e+02   
50%                 3.970000             3.960000  4.403000e+03   
75%                 4.120000             4.140000  2.014300e+04   
max                 5.000000             5.000000  3.820921e+06   

         num_reviews          score  Frequency     new_score   super_score  
count   22891.000000   22891.000000        0.0  22891.000000  22891.000000  
mean     2324.754794    3893.028483        NaN      0.032536      2.307648  
std      6837.503574   11022.018268        NaN      0.092116      0.156127  
min         0.000000      55.000000        NaN      0.000460      0.871881  
25%       106.000000     832.000000        NaN      0.006953      2.211681  
50%       384.000000    1727.000000        NaN      0.014433      2.310253  
75%      1504.000000    3597.500000        NaN      0.030066      2.406540  
max    147696.000000  598270.000000        NaN      5.000000      4.625000

### Problem 9
Use super_score column to find out how many unique authors have a super_score of 3.9 or more. Let's call these authors, super_authors. Create a dataframe called super_authors that has the following columns: ['author_name','author_gender','birthplace','book_title'] and show the head of the data.

In [69]:
books1.loc[(books1["super_score"] >= 3.9)]

author_average_rating author_gender         author_name  \
919                   4.09          male       Oscar Wilde\n   
920                   4.06          male     Lewis Carroll\n   
921                   4.33          male  Shel Silverstein\n   

                birthplace  book_average_rating  \
919          Ireland\n                     4.07   
920   United Kingdom\n                     4.07   
921    United States\n                     4.37   

                                            book_title    genre_1  \
919               \n      The Picture of Dorian Gray\n   Classics   
920  \n      Alice's Adventures in Wonderland & Thr...   Classics   
921                          \n      The Giving Tree\n  Childrens   

     num_ratings  num_reviews pages   score  Frequency  new_score  super_score  
919       793936        23624   367  442035        NaN   3.694277     3.916996  
920       414440         9294   239  489613        NaN   4.091907     4.076801  
921       801232        15919    64  598270        NaN   5.000000     4.625000

In [79]:
authors = pd.DataFrame({'author_name':books1['author_name'], 'author_gender':books1['author_gender'], 
                              'birthplace':books1['birthplace'], 'book_title':books1['book_title']})

In [80]:
super_authors = authors.loc[(books1["super_score"] >= 3.9)]

In [81]:
super_authors.head()

author_name author_gender             birthplace  \
919       Oscar Wilde\n          male          Ireland\n       
920     Lewis Carroll\n          male   United Kingdom\n       
921  Shel Silverstein\n          male    United States\n       

                                            book_title  
919               \n      The Picture of Dorian Gray\n  
920  \n      Alice's Adventures in Wonderland & Thr...  
921                          \n      The Giving Tree\n

### Problem 10
Save the dataframe super_authors you created in problem 9 to your working folder (where you have this HW saved) as a csv file and read it back to Jupyter Notebook and print the head of the data.

In [82]:
super_authors.to_csv('super_authors.csv')

In [83]:
superauthors = pd.read_csv('super_authors.csv')

In [84]:
superauthors.head()

Unnamed: 0         author_name author_gender             birthplace  \
0         919       Oscar Wilde\n          male          Ireland\n       
1         920     Lewis Carroll\n          male   United Kingdom\n       
2         921  Shel Silverstein\n          male    United States\n       

                                          book_title  
0               \n      The Picture of Dorian Gray\n  
1  \n      Alice's Adventures in Wonderland & Thr...  
2                          \n      The Giving Tree\n

## The following 2 problems are extra credit problems and are optional. You won't get any help on these problems.

### Bonus Problem 1 (Bonus 5 points)
In super_author data set, there are a few columns that have entries with $\text{\n}$ notation. Find these columns and remove $\text{\n}$ from all the column entries using regular expressions.  Read more about handling strings and regular expressions here. https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html, https://www.geeksforgeeks.org/replace-values-in-pandas-dataframe-using-regex/
, and here https://docs.python.org/2/library/re.html.

In [86]:
authors_updated = super_authors.replace(to_replace ='\n', value = '', regex = True)

In [87]:
authors_updated.head()

author_name author_gender           birthplace  \
919       Oscar Wilde          male          Ireland       
920     Lewis Carroll          male   United Kingdom       
921  Shel Silverstein          male    United States       

                                            book_title  
919                         The Picture of Dorian Gray  
920        Alice's Adventures in Wonderland & Throu...  
921                                    The Giving Tree

### Bonus Problem 2 (Bonus 5 points)
You may have noticed that the column named  "pages" from the data set in problem 1 seem to have numerical looking values but it is not showing up in our data summary. This is because this columns appears to be a string in our data set. Data values in this columns are just  the total number of pages in each book and these should be numbers not strings. Change this column from string to numeric and find mean, median and mode for this column. If the error comes in the process of converting to numerical, find out what the error is and fix it.

In [90]:
books1["pages"] = pd.to_numeric(books1["pages"], errors='coerce')

In [91]:
books1.dtypes

author_average_rating    float64
author_gender             object
author_name               object
birthplace                object
book_average_rating      float64
book_title                object
genre_1                   object
num_ratings                int64
num_reviews                int64
pages                    float64
score                      int64
Frequency                float64
new_score                float64
super_score              float64
dtype: object

In [92]:
books1['pages'].mean()

332.99602411744144

In [93]:
books1['pages'].median()

315.0

In [94]:
books1['pages'].mode()

0    320.0
dtype: float64